### DISCUSSION TITLE
Is the transaction context the same for any contract?

### DISCUSSION THREAD

**Discussion by ArtyomVancyan** [hyperledger/fabric#4198] [Q&A] [answered]
We have an issue on our ledger where we have two types of contract entities whose keys can cross (be the same) in the world state. And creating an entity with the same key after another, overwrites the first one. Our contracts have a structure similar to the following.

```go
package chaincode

import (
	"github.com/hyperledger/fabric-contract-api-go/contractapi"
)

type CredentialContract struct {
	contractapi.Contract
}

func (s *CredentialContract) GetCredentialById(ctx contractapi.TransactionContextInterface, credentialId string) ([]byte, error) {
	credentialJSON, err := ctx.GetStub().GetState(credentialId)
	if err != nil {
		return nil, err
	}
	return credentialJSON, nil
}
```

And in the main package, we initialize them in the following way:

```go
package main

import (
	"log"

	"ledger-service/chaincode"
	"github.com/hyperledger/fabric-contract-api-go/contractapi"
)

func main() {
	// Initialize Smart Contracts
	contracts, err := contractapi.NewChaincode(
		&chaincode.CredentialContract{},
	if err != nil {
		log.Panicf("Error creating smart contracts: %v", err)
	}

	if err := contracts.Start(); err != nil {
		log.Panicf("Error starting smart contracts : %v", err)
	}
}
```

Was not the context (a.k.a: `ctx` in the code) supposed to be private per contract? How do I make it to be private per contract and not share the same state?

**Comment by bestbeforetoday** [comment, marked as answer]
Ledger keys written by a smart contract are scoped (or namespaced) within the channel's ledger by the name of their containing chaincode. Smart contracts packaged within a single chaincode share the same chaincode namespace. You need to either ensure that smart contracts within the same chaincode package don't inadvertently use the same ledger keys, or split your smart contracts into separate chaincode packages so their keys are contained within separate namespaces.

**Reply by ArtyomVancyan** [reply]
> Smart contracts packaged within a single chaincode share the same chaincode namespace.

This is where my problem comes from. Thank you a lot for the helpful answer,
